File: src/main/python/leetcode/tree/path_sum_two.py
from .models import TreeNode
from typings import List

class Solution:

    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        self.target = sum
        self.ret = []
        self.traverse(root, [], 0)
        return self.ret


    def traverse(self, node: TreeNode, l: List[int], cur_sum: int):
        if node is None:
            return

        l.append(node.val)

        if node.left is None and node.right is None and (node.val + cur_sum == self.target):
            self.ret.append(l.copy())
        else:
            self.traverse(node.left, l, cur_sum + node.val)
            self.traverse(node.right, l, cur_sum + node.val)

        l.pop()

File: src/main/java/leetcode/tree/PathSumTwo.java
package leetcode.tree;

import java.util.*;

public class PathSumTwo {
  List<List<Integer>> ret = new ArrayList<>();
  int target;

  public List<List<Integer>> pathSum(TreeNode root, int sum) {
    this.target = sum;
    traverse(root, new ArrayList<>(), 0);
    return ret;
  }

  void traverse(TreeNode root, List<Integer> list, int curSum) {

    if (root == null) {
      return;
    }

    list.add(root.val);

    if (root.left == null && root.right == null && root.val + curSum == this.target) {
      ret.add(list);
    }

    traverse(root.left, new ArrayList(list), curSum + root.val);
    traverse(root.right, new ArrayList(list), curSum + root.val);

  }
}
